### DISCUSSION TITLE
How to add aliases in php-fpm shell?

### DISCUSSION THREAD

**Discussion by MagePsycho** [wardenenv/warden#564] [Help & Support]
I would like to add some aliases on `php-fpm` shell, for example:
```
alias m2="./bin/magento"
```

I could add it `~/.bashrc` but it would be gone in re-build.

Is there a way we can add it via `.warden/warden-env.yml` somehow? or can we add the `m2` alias on warden level?

**Comment by tdgroot** [comment]
@MagePsycho you could add alias suggestions to the image, and otherwise you could add aliases to your host:
``` shell
alias m2="warden env-exec php-fpm bin/magento"
```

You could also write a bash function on your host which copies your preferred `.bash_profile` (or whatever) to the container before it executes `warden shell`.

**Reply by MagePsycho** [reply]
> you could add alias suggestions to the image

This is what I am looking for. Can you guide me on how to add an alias to the image itself?

**Comment by MagePsycho** [comment]
@tdgroot I have already created some aliases for `warden` commands. For example:

## Aliases
With aliases, you don't need to write full `warden` command.

| Category | Warden Command | Alias Command |
| ------- | -------------- | ------------ |
| Magento 2   | `warden shell -c "bin/magento <action>"`  | `wm2 <action>` |
| MageRun   | `warden shell -c "n98-magerun <action>"` | `wmr <action>` |
| Redis   | `warden redis <action>` | `wrd <action>` |
| Database  | `warden db <action>` | `wdb <action>` |
| Logs | `warden env logs --tail 0 -f nginx php-fpm php-debug`  | `wlg` |
| Shell (`php-fpm`) | `warden shell`  | `wsh` |
| Shell (`php-debug`) | `warden debug`  | `wdg` |
| Environment | `warden env <action>`  | `wenv <action>` |

Source - https://github.com/MagePsycho/magento2-warden-aliases

But I would like to add some aliases in the container itself.

**Comment by ihor-sviziev** [comment]
@davidalger, what do you think about adding these aliases to the warden itself?

**Comment by ihor-sviziev** [comment]
Let's keep it open

**Comment by MagePsycho** [comment]
Is there a way to add the following alias **permanently**
```
alias m2="./bin/magento"
```

So that we can run any mage commands via
```
m2 <comand>
```

I can see that I can use MageRun alias in the shell as:
```
mr <command>
```
I want something similar for `bin/magento`

**Reply by navarr** [reply]
The easiest way at current would be to mount a .bashrc file via `.warden/warden-env.yml` to the php-fpm and php-debug services.

After the Den changes are merged in, I believe there will be a way to do this once and have it apply to all projects - but I'll have to double check that.

**Comment by MagePsycho** [comment]
@navarr 
How to mount a `.bashrc` via `.warden/warden-env.yml`?

Also, which Den changers are you referring to?

**Reply by MagePsycho** [reply]
For the mounting part, I got it working.
File: `.warden/warden-env.yml`
```
version: "3.5"
services:
  php-fpm:
    volumes:
      - ./.warden/php-fpm/.bashrc:/home/www-data/.bashrc
```

Where content of file: `./.warden/php-fpm/.bashrc` looks like this
```	  
# .bashrc

# Source global definitions
if [ -f /etc/bashrc ]; then
	. /etc/bashrc
fi

# User specific environment
if ! [[ "$PATH" =~ "$HOME/.local/bin:$HOME/bin:" ]]
then
    PATH="$HOME/.local/bin:$HOME/bin:$PATH"
fi
export PATH

# Uncomment the following line if you don't like systemctl's auto-paging feature:
# export SYSTEMD_PAGER=

# User specific aliases and functions	  
alias m2=./bin/magento
```

Is there a way to make it global? or add it by default `m1` for Magento 1 and `m2` for Magento 2 (like warden has added `mr` by default for MageRun)

**Reply by bap14** [reply]
You are more than welcome to fork your own base images and add these aliases, then in your project's `.warden/warden-env.yaml` file specify the new image to use. The downside being that you'd have to rebuild and tag those images. But that's an immediate way to add the aliases directly into the image.

**Comment by bap14** [comment]
The issue I see with baking in aliases like this is that not every alias will always be available. When using existing sources that alias will point to an executable that doesn't exist until the user runs `composer install`. Tools like n98-magerun are technically "always available" and so a baked in alias like `mr` makes sense.

One solution I think would solve this, or at least make it easier, would be to customize the shell initialization script (e.g. `.bashrc`) so that it looks at a directory for extra custom initialization scripts. I could see a directory like `$HOME/bashrc.d` being baked into the image, and during startup `$WARDEN_HOME_DIR/bashrc.d` is iterated over to mount individual files to `$HOME/bashrc.d`. Then we'd do the same thing for a project's `$WARDEN_ENV_PATH/.warden/bashrc.d`. This way a project can still override the global initialization script(s).

Now where this could get hairy is in multiple shells. I know Den added support for Zsh and Fish. Zsh should be able to just consume the same init scripts as bash, but Fish has a different methodology (actually it's already built in if you just mount files to `$HOME/.config/fish/conf.d/`). So we may need to specify that Bash/Zsh scripts end in `.sh` and Fish scripts end in `.fish` (as required) or use subdirectories like `.warden/init-scripts/bash_zsh/` and `.warden/init-scripts/fish`.

Would something like the above be acceptable @MagePsycho ? Thoughts @navarr ?

**Reply by MagePsycho** [reply]
I still feel the use of `m2` in the same level of `mr`. If the project type is Magento2 then create an alias during spinning the container.

**Reply by MagePsycho** [reply]
Still wondering, if there is a way of adding aliases globally for all the warden projects.
This would be very helpful in reusing your local bash aliases in all warden projects.

CC: @bap14
